Reorganize rpc API doc and add introduction

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30516 Add examples to RRef doc
* #30515 Make doc source format consistent in rpc/init.cpp
* **#30491 Reorganize rpc API doc and add introduction**

Our RPC API docs presents the APIs well but misses a general
introduction to the APIs. Readers might be a little lost the first
time landing this page. This commits reorganizes the APIs into
four components from user's perspective, RPC, RRef, dist autograd,
and dist optimizer. It also adds an intro to each and briefly
discribes why we provide those.

Differential Revision: [D18723294](https://our.internmc.facebook.com/intern/diff/D18723294)

mrshenli committed:
[WIP] Reorganize rpc API doc

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Reorganize rpc API doc"




[ghstack-poisoned]

mrshenli committed:
Update on "Reorganize rpc API doc and add introduction"


Our RPC API docs presents the APIs well but misses a general
introduction to the APIs. Readers might be a little lost the first
time landing this page. This commits reorganizes the APIs into
four components from user's perspective, RPC, RRef, dist autograd,
and dist optimizer. It also adds an intro to each and briefly
discribes why we provide those.

Differential Revision: [D18723294](https://our.internmc.facebook.com/intern/diff/D18723294)

[ghstack-poisoned]

rohan-varma commented on docs/source/rpc.rst:
```diff
@@ -11,8 +11,65 @@ machines.
+   be launched, which can become quite complicated where there are nested and
+   inter-dependent RPC calls in the forward pass.
+4) **Distributed Optimizer** takes a list of parameter RRefs in the constructor
+   and updates them all accordingly when running `step()`. When you have a
```
Mention that the distributed optimizer takes in a regular optimizer, and runs the step() function of that (so it can do SGD, Adagrad, etc)

rohan-varma commented on docs/source/rpc.rst:
```diff
@@ -11,8 +11,65 @@ machines.
+   inter-dependent RPC calls in the forward pass.
+4) **Distributed Optimizer** takes a list of parameter RRefs in the constructor
+   and updates them all accordingly when running `step()`. When you have a
+   distributed forward and backward passes, parameters and gradients will be
```
Nit: change "when you have a distributed forward and backward passes" to "when you have distributed forward and backward passes" (remove the "a")

rohan-varma reviewed (approved):
Overall looks good % a couple inline comments. Thanks!

mrshenli committed:
Update on "Reorganize rpc API doc and add introduction"


Our RPC API docs presents the APIs well but misses a general
introduction to the APIs. Readers might be a little lost the first
time landing this page. This commits reorganizes the APIs into
four components from user's perspective, RPC, RRef, dist autograd,
and dist optimizer. It also adds an intro to each and briefly
discribes why we provide those.

Differential Revision: [D18723294](https://our.internmc.facebook.com/intern/diff/D18723294)

[ghstack-poisoned]
